opal/runtime/opal_params_core.c

```diff
@@ -350,3 +422,73 @@ int opal_register_util_params(void)
 
     return OPAL_SUCCESS;
 }
+
+/* Parses a color 'string' and extracts the 'code' for each 'key'
+ * The string is in the format of 'key1=code1:key2=code2:...'
+ * Example: opal_var_dump_color MCA parameter
+ *
+ * color_string: The string containing the key=code sequences to parse
+ * key_names: The keys to extract (e.g. key1, key2)
+ * key_count: The length of key_names
+ * values_out: The extracted values for the keys, 1-1 with key_names
+ *   ATTENTION: Make sure it's at least as large as key_names...
+ */
+static int parse_color_string(char *color_string, char **key_names,
+        int key_count, char **values_out) {
+
+    char **tokens = opal_argv_split(color_string, ':');
+    if (NULL == tokens) {
+        return OPAL_ERR_OUT_OF_RESOURCE;
+    }
+
+    for (int i = 0; tokens[i] != NULL; i++) {
+        char **kv = opal_argv_split(tokens[i], '=');
+
+        if (NULL == kv) {
+            opal_argv_free(tokens);
+            return OPAL_ERR_OUT_OF_RESOURCE;
+        }
+
+        // Expected format of token: key=code
+        if (opal_argv_count(kv) != 2) {
+            opal_argv_free(kv);
+            continue;
+        }
+
+        // Look for key name and store value in respective position
+        for (int k = 0; k < key_count; ++k) {
+            if (strcasecmp(key_names[k], kv[0]) == 0) {
+                int ret = opal_asprintf(&values_out[k], "\033[%sm", kv[1]);
+
+                if (ret < 0) {
+                    opal_argv_free(kv);
+                    opal_argv_free(tokens);
+                    return OPAL_ERR_OUT_OF_RESOURCE;
+                }
+
+                break;
+            }
+        }
+
+        opal_argv_free(kv);
+    }
+
+    opal_argv_free(tokens);
+
+    for (int k = 0; k < key_count; k++) {
+        if (NULL == values_out[k]) {
+            values_out[k] = strdup(""); // needs to be free-able
+
+            if (NULL == values_out[k]) {
+                for (int v = 0; v < k; v++) {
+                    free(values_out[v]);
+                    values_out[v] = NULL;
+                }
```

Comment:
I think you can just `opal_argv_free(&values_out)` here...?

Might want to check the overall lifecycle of `opal_var_dump_color` -- both in the error and no-error cases to make sure everything gets freed properly.

Reply:
Looks like it would work due to `values_out[k] == NULL`, but feels hacky since values_out is generally not NULL-terminated.

Generally (in the non-error case) the values are freed in `opal_deregister_util_params` -- I'm not entirely sure at exactly which stage of finalize this gets called. Another alternative could be to free them in `mca_base_var_finalize` (?). In the error case (`OPAL_ERR_OUT_OF_RESOURCE`), theoretically all values are freed, and I expect that the error will propagated and OMPI will fail (and thus `opal_var_dump_color`'s NULL values won't get used in `mca_base_var*`).

Reply:
> Looks like it would work due to `values_out[k] == NULL`, but feels hacky since values_out is generally not NULL-terminated.

Fair enough.

> Generally (in the non-error case) `opal_var_dump_color` is freed in `opal_deregister_util_params` -- I'm not entirely sure at exactly which stage of finalize this gets called. Another alternative could be to free them in `mca_base_var_finalize` (?). 

That's probably good enough; just as long as they are actually cleaned up properly in the non-error case.

> In the error case (`OPAL_ERR_OUT_OF_RESOURCE`), theoretically all values are freed, and I expect that the error will propagated and OMPI will fail (and thus `opal_var_dump_color`'s NULL values won't get used in `mca_base_var*`).

👍